pkg/clients/dclient/discovery.go

```diff
@@ -92,88 +93,280 @@ func (c serverPreferredResources) GetGVRFromAPIVersionKind(apiVersion string, ki
-			return r, gvr, nil
+		if gvk, err := c.findResourceFromResourceName(groupVersion, resourceName); err == nil {
+			return gvk, nil
 		}
 	}
 
-	return nil, schema.GroupVersionResource{}, err
+	return schema.GroupVersionKind{}, err
 }
 
-func (c serverPreferredResources) findResource(apiVersion string, kind string) (*metav1.APIResource, schema.GroupVersionResource, error) {
-	var serverResources []*metav1.APIResourceList
-	var err error
-	if apiVersion == "" {
-		serverResources, err = c.cachedClient.ServerPreferredResources()
-	} else {
-		_, serverResources, err = c.cachedClient.ServerGroupsAndResources()
+// FindResource finds an API resource that matches 'kind'. For finding subresources that have the same kind as the parent
+// resource, kind has to be specified as 'ParentKind/SubresourceName'. For matching status subresource of Pod, kind has
+// to be specified as `Pod/status`. If the resource is not found and the Cache is not fresh, the cache is invalidated
+// and a retry is attempted
+func (c serverResources) FindResource(groupVersion string, kind string) (apiResource, parentAPIResource *metav1.APIResource, gvr schema.GroupVersionResource, err error) {
+	r, pr, gvr, err := c.findResource(groupVersion, kind)
+	if err == nil {
+		return r, pr, gvr, nil
+	}
+
+	if !c.cachedClient.Fresh() {
+		c.cachedClient.Invalidate()
+		if r, pr, gvr, err = c.findResource(groupVersion, kind); err == nil {
+			return r, pr, gvr, nil
+		}
+	}
+
+	return nil, nil, schema.GroupVersionResource{}, err
+}
+
+// findResourceFromResourceName returns the GVK for the a particular resourceName and groupVersion
+func (c serverResources) findResourceFromResourceName(groupVersion, resourceName string) (schema.GroupVersionKind, error) {
+	_, serverGroupsAndResources, err := c.cachedClient.ServerGroupsAndResources()
+	if err != nil && !strings.Contains(err.Error(), "Got empty response for") {
+		if discovery.IsGroupDiscoveryFailedError(err) {
+			logDiscoveryErrors(err)
+		} else if isMetricsServerUnavailable(groupVersion, err) {
+			logger.V(3).Info("failed to find preferred resource version", "error", err.Error())
+		} else {
+			logger.Error(err, "failed to find preferred resource version")
+			return schema.GroupVersionKind{}, err
+		}
+	}
+	return findResourceFromResourceName(groupVersion, resourceName, serverGroupsAndResources)
+}
+
+func findResourceFromResourceName(groupVersion string, resourceName string, serverGroupsAndResources []*metav1.APIResourceList) (schema.GroupVersionKind, error) {
+	for _, serverResourceList := range serverGroupsAndResources {
+		for _, apiResource := range serverResourceList.APIResources {
+			serverResourceGroupVersion := getServerResourceGroupVersion(serverResourceList.GroupVersion, apiResource.Group, apiResource.Version)
+			if serverResourceGroupVersion == groupVersion && apiResource.Name == resourceName {
+				logger.V(4).Info("found preferred resource", "groupVersion", groupVersion, "resourceName", resourceName)
```

Comment:
Suggest changing it to log level 6.

Reply:
I changed logging level to 6 for:
```
logger.V(6).Info("gv with resource", "gvWithResource", groupVersionResource)
```
but have kept it 4 for:
```
logger.V(4).Info("matched API resource to kind", "apiResource", serverResource, "kind", kind)
```
As earlier also it was 4, which is fine imo.

Reply:
I found it quite annoying while debugging at level 4, also changed it to level 6 in another PR :)

https://github.com/kyverno/kyverno/pull/5409/files#diff-0fcb52bb9daafa34f23e4492cfd88eb277bd43a3915b4ed191d94e0a2f7f621bR149

Reply:
Changed to 6, done.